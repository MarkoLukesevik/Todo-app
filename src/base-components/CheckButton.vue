<script setup lang="ts">
import { computed } from 'vue'

const props = defineProps<{
  isChecked: boolean
}>()

const emit = defineEmits<{
  (e: 'toggle', value: boolean): void
}>()

const checkButtonBackground = computed<string>(() => {
  return props.isChecked ? 'linear-gradient(hsl(192, 100%, 67%), hsl(280, 87%, 65%))' : ''
})
</script>

<template>
  <button
    class="check-button"
    :style="{ background: checkButtonBackground }"
    @click="emit('toggle', !isChecked)"
  >
    <img v-if="isChecked" src="/assets/images/icon-check.svg" alt="" />
  </button>
</template>

<style scoped lang="scss">
.check-button {
  display: flex;
  flex-direction: row;
  flex: 0 1 auto;
  align-items: center;
  justify-content: center;

  width: 20px;
  height: 20px;

  background-color: transparent;
  border: 1px solid #979797;
  border-radius: 50%;

  &:hover {
    background:
      linear-gradient(white, white) padding-box,
      linear-gradient(to right, hsl(192, 100%, 67%), hsl(280, 87%, 65%)) border-box;
    border: 1px solid transparent;
  }
}
</style>
