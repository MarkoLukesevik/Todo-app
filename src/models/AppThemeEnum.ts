export enum AppThemeEnum {
    Light = 0,
    Dark = 1
}
