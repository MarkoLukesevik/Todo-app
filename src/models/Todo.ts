export interface Todo {
  id: string
  title: string
  isComplete: false
}
