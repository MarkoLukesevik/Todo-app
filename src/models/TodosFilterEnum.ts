export enum TodosFilterEnum {
    All = 0,
    Active = 1,
    Completed = 2
}