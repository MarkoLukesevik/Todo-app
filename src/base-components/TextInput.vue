<script setup lang="ts">
import { AppThemeEnum } from '@/models/AppThemeEnum';
import { computed } from 'vue';

const props = defineProps<{
    placeholder: string;
    modelValue: string;
    appTheme: AppThemeEnum;
}>();

const emit = defineEmits<{
    (e: 'update:modelValue', value: string): void;
}>();

const textInputColor = computed<string>(() => {
    return props.appTheme === AppThemeEnum.Light ? '#9495a5' : '#767992';
});

const handleInputChange = (event: Event): void => {
    emit('update:modelValue', (event.target as HTMLInputElement).value);
};
</script>

<template>
  <input
    class="app-input"
    :style="{ color: textInputColor }"
    :placeholder="placeholder"
    :value="modelValue"
    @input="handleInputChange($event)"
  />
</template>

<style scoped lang="scss">
.app-input {
    display: flex;
    flex-direction: row;
    flex: 1 1 auto;

    background-color: transparent;
    border: none;

    font-family: 'Josefin Sans', sans-serif;
    font-size: 0.9rem;
    font-weight: 400;

    &:focus {
        outline: none;
    }
}
</style>
