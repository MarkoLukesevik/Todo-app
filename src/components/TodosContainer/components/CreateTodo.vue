<script setup lang="ts">
import { ref, computed } from 'vue';
import type { Ref } from 'vue';

import { AppThemeEnum } from '@/models/AppThemeEnum';
import type { Todo } from '@/models/Todo';
import { generateId } from '@/helpers/generateid';

import CheckButton from '@/base-components/CheckButton.vue';
import TextInput from '@/base-components/TextInput.vue';

const props = defineProps<{
    appTheme: AppThemeEnum;
}>();

const emit = defineEmits<{
  (e: 'create-todo', value: Todo): void;
}>();

const newTodoTitle: Ref<string> = ref('');
const newTodoIsComplete: Ref<boolean> = ref(false);

const containerBackgroundColor = computed<string>(() => {
    return props.appTheme === AppThemeEnum.Light ? '#ffffff' : '#25273D';
});

const handleAddNewTodo = (): void => {
    if (!newTodoTitle.value) return;

    emit('create-todo', {
        id: generateId(),
        title: newTodoTitle.value,
        isComplete: newTodoIsComplete.value
    });

    resetNewTodo();
}

const resetNewTodo = (): void => {
    newTodoIsComplete.value = false;
    newTodoTitle.value = '';
}

const toggleNewTodo = (): void => {
    newTodoIsComplete.value = !newTodoIsComplete.value;
}
</script>

<template>
    <div
        data-test="create-todo-container"
        class="create-todo-container"
        @keyup.enter="handleAddNewTodo()"
        :style="{ backgroundColor: containerBackgroundColor }"
    >
        <check-button
            :isChecked="newTodoIsComplete"
            :app-theme="appTheme"
            @toggle="toggleNewTodo"
        />
        <text-input
            placeholder="Create new todo..."
            v-model="newTodoTitle"
            :app-theme="appTheme"
        />
    </div>
</template>

<style scoped lang="scss">
.create-todo-container {
    display: flex;
    flex: 0 1 auto;
    flex-direction: row;
    column-gap: 0.7rem;

    padding: 1rem;
    border-radius: 5px;
    background-color: white;
}
</style>
