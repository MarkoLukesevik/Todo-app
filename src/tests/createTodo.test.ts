import { beforeEach, describe, expect, it, vi } from "vitest";
import { shallowMount } from "@vue/test-utils";

import { AppThemeEnum } from "@/models/AppThemeEnum";

import CreateTodo from "@/components/TodosContainer/components/CreateTodo.vue";
import CheckButton from "@/base-components/CheckButton.vue";
import TextInput from "@/base-components/TextInput.vue";

vi.mock('@/helpers/generateid', () => ({
    generateId: vi.fn(() => '12345'),
}));

describe('Test suite for create todo component', () => {
    let wrapper: any;

    beforeEach(() => {
        wrapper = shallowMount(CreateTodo, {
            props: {
                appTheme: AppThemeEnum.Light
            }
        });
    });

    it('Should match the snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    it('Should render correct background color based on appTheme prop', async () => {
        const container = await wrapper.find('[data-test="create-todo-container"]');

        expect(container.element.style.getPropertyValue('background-color')).toBe('rgb(255, 255, 255)');

        await wrapper.setProps({ appTheme: AppThemeEnum.Dark });
        expect(container.element.style.getPropertyValue('background-color')).toBe('rgb(37, 39, 61)');
    });

    it('Should emit create-todo event with correct data when handleAddNewTodo is called', async () => {
        wrapper.vm.newTodoTitle = 'Test Todo';
        wrapper.vm.newTodoIsComplete = true;

        await wrapper.vm.handleAddNewTodo();

        expect(wrapper.emitted('create-todo')).toHaveLength(1);
        expect(wrapper.emitted('create-todo')[0]).toEqual([{
            id: '12345',
            title: 'Test Todo',
            isComplete: true,
        }]);

        expect(wrapper.vm.newTodoTitle).toBe('');
        expect(wrapper.vm.newTodoIsComplete).toBe(false);
    });

    it('Should toggle newTodoIsComplete when toggleNewTodo is called', async () => {
        expect(wrapper.vm.newTodoIsComplete).toBe(false);

        await wrapper.vm.toggleNewTodo();
        expect(wrapper.vm.newTodoIsComplete).toBe(true);

        await wrapper.vm.toggleNewTodo();
        expect(wrapper.vm.newTodoIsComplete).toBe(false);
    });

    it('triggers handleAddNewTodo when Enter key is pressed', async () => {
        wrapper.vm.newTodoTitle = 'Test Todo';
    
        await wrapper.trigger('keyup.enter');
    
        expect(wrapper.emitted('create-todo')).toHaveLength(1);
        expect(wrapper.emitted('create-todo')[0]).toEqual([{
            id: '12345',
            title: 'Test Todo',
            isComplete: false,
        }]);
    });

    it('Should render check-button and text-input components', () => {
        const checkButton = wrapper.findComponent(CheckButton);
        expect(checkButton.exists()).toBe(true);
    
        const textInput = wrapper.findComponent(TextInput);
        expect(textInput.exists()).toBe(true);
    });
});