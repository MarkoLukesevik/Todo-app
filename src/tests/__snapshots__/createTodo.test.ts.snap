// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test suite for create todo component > Should match the snapshot 1`] = `
<div
  class="create-todo-container"
  data-test="create-todo-container"
  data-v-a79cb070=""
  style="background-color: rgb(255, 255, 255);"
>
  <check-button-stub
    apptheme="0"
    data-v-a79cb070=""
    ischecked="false"
  />
  <text-input-stub
    apptheme="0"
    data-v-a79cb070=""
    modelvalue=""
    placeholder="Create new todo..."
  />
</div>
`;
